keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1096 @@
+
+To address the PreBind concern, the solution is to modify the scheduling framework to flush the in-flight events cache before PreBind.
+This prevents issues in the scheduling queue caused by keeping pods at PreBind for an extended period.
+This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.
+This issue needs to be resolved before the beta is released.
+
+## Design Details
+
+### DRA Scheduler Plugin Design Overview
+
+This document outlines the design of the DRA Scheduler Plugin, focusing on the handling of fabric devices.
+Key additions include `BindingConditions` and `BindingFailureConditions for device identification and preparetion, enhancements to `AllocatedDeviceStatus.Conditions`, and the process for handling `ResourceSlices` upon attachment failure.
+The composable controller design is also discussed, emphasizing efficient utilization of fabric devices.
+
+![proposal](proposal.jpg)
+
+#### BasicDevice Enhancements
+
+To indicate that the device is a Fabric device or other device that requires some preparation (e.g. attachment), some fields are added to the `Basic` within `Device`.
+These fields will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
+
+```go
+// BasicDevice represents a basic device instance.
+type BasicDevice struct {
+    // BindingConditions defines the conditions for binding.
+    //
+    // +optional
+    BindingConditions []string
+
+    // BindingFailureConditions defines the conditions for binding failure.
+    //
+    // +optional
+    BindingFailureConditions []string
+
+    // UsageRestrictedToNode indicates if the usage of an allocation involving this device
+    // has to be limited to exactly the node that was chosen when allocating the claim.
+    //
+    // +optional
+    UsageRestrictedToNode bool
+
+    // BindingTimeoutSeconds indicates the prepare timeout period.
+    // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
+    //
+    // +optional
+    BindingTimeoutSeconds *metav1.Duration
+}
+
+const (
+    BindingConditionsMaxSize    = 4
+    BindingFailureConditionsMaxSize = 4
+)
+```
+
+#### AllocatedDeviceStatus.Conditions Enhancements
+
+The `BindingConditions` and `BindingFailureConditions` fields within `AllocatedDeviceStatus.Conditions` are used to indicate the status of the device attachment.
+These fields will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, following fields are added:
+
+```go
+// AllocatedDeviceStatus.Conditions contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus.Conditions struct {
+    ...
+    // BindingConditions defines the conditions for binding.
+    //
+    // +optional
+    BindingConditions map[string]bool
```

Comment:
We lost @pohly's previous comments here. This is here to capture the "as-was" values for these fields - the values as they were set in the ResourceSlice at the time of allocation. I think this should be `[]string`, and we need to copy over the timeout as well, and maybe the UsageRestrictedToNode?